### PR TITLE
DBZ-4243

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.redhat.com/browse/DBZ-4243

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Supersedes https://github.com/debezium/debezium/pull/2893/.

**Comment by gunnarmorling** [comment]
@jpechane, so one of the new tests is failing: `AbstractIncrementalSnapshotWithSchemaChangesSupportTest::renameTable()`. The name is confusing, as it actually adds a table column. The issue is that the newly added column (`c`) isn't part of the snapshot select. I'm wondering about the intended timing here: it seems we'd have to re-create the snapshot select for each chunk, so to accomodate for any concurrently happening schema changes. Would you agree? What about the isolation level, though? I take it that any concurrent schema change won't impact a currently ongoing chunk TX, right?

**Comment by Naros** [comment]
> I take it that any concurrent schema change won't impact a currently ongoing chunk TX, right?

At least for PostgreSQL, I stumbled onto a similar problem in PR #2925 where the PostgreSQL driver caches some schema data and when using the default autosave mode of `never`, an error is thrown when dropping a column.

**Comment by gunnarmorling** [comment]
Ok, I think there may be a more general problem here. @jpechane, @Naros, let's try and have a discussion about it some time tomorrow.

**Comment by gunnarmorling** [comment]
>  default autosave mode

@Naros, what is this btw.?

**Comment by Naros** [comment]
See https://jdbc.postgresql.org/documentation/head/connect.html

**autosave** = String
_Specifies what the driver should do if a query fails. In autosave=always mode, JDBC driver sets a savepoint before each query, and rolls back to that savepoint in case of failure. In autosave=never mode (default), no savepoint dance is made ever. In autosave=conservative mode, savepoint is set for each query, however the rollback is done only for rare cases like 'cached statement cannot change return type' or 'statement XXX is not valid' so JDBC driver rollsback and retries_

_Default is `never`_.

So basically when you have a cached query and the table's structure changes, if you don't explicitly specify a value other than `never` then you'll end up with the driver throw an error that says `ERROR:  cached plan must not change result type`.

**Comment by dude0001** [comment]
Is there any chance this still makes it into 1.8.0.CR1 (per the status in the issue tracker)? If the refactor is blocking, could we still consider making my original change for this a pure bug fix and spinning the refactor off into a new issue? We really need this for a go-live and is currently a blocker for us. I'm willing to help as well, but not sure where this is blocked currently.

**Comment by gunnarmorling** [comment]
Hey @dude0001, sorry about the radio silence on this one, this feel a bit through the cracks, with all the other stuff going on :(

In terms of where this is blocked, the issue is DDL changes happening while an incremental snapshot is running. By selecting only the filter-included columns, this causes some tests to fail as of this PR. I think the right way to handle this will be to re-create the snapshot SELECT for each chunk transaction as per the then-valid schema of that table. Any help with doing that would be more than welcomed.

Regarding a timeline, we are planning to do the 1.8 Final, and even if we had that change ready right now, unfortunately I don't see us adding it that late in the release cycle. So this would have to wait for 1.8.1.Final, due some time in January.

**Comment by dude0001** [comment]
We are dealing with Log4J and other issues as well. It is a difficult situation for a lot of people. I apologize and did not mean to but any extra pressure on you @gunnarmorling. This is an excellent community and we've had nothing but a great experience with this project, even when we run into issues the help is great. Thank you for the update and keep up the great work!

**Comment by gunnarmorling** [comment]
Hey @jpechane, could you let me know what you think about this one? It is failing a test (DDL change while an incremental snapshot is running), I still believe though it represents an improvement over the status quo. I.e. I'd value the ability to incrementally snapshot tables with only some columns CDC-enabled higher than incrementally snapshotting while a DDL schema change is happening (one should be able to prevent that situation by procedural means quite easily). WDYT?

**Comment by jpechane** [comment]
@gunnarmorling I'll give you the answer later on. I'm in the process of rebasing the PR on the latest main i is quite convoluted process. When done, I'll recheck the results.

**Comment by gunnarmorling** [comment]
Closing in favour of #3320. It seems @jpechane has a fix handy for that schema change issue.
